corehq/messaging/util.py

```diff
@@ -41,14 +49,26 @@ def rule_initiation_key_is_set(self):
     def rule_initiation_key_minutes_remaining(self):
         return (self.client.ttl(self.rule_initiation_key) // 60) or 1
 
-    def set_initial_progress(self):
+    def set_initial_progress(self, shard_count=0):
+        # shard_count is passed when tasks run pear each shard
         self.client.set(self.current_key, 0)
         self.client.set(self.total_key, 0)
+        if shard_count:
+            self.client.set(self.shard_count_key, shard_count)
         self.client.expire(self.current_key, self.key_expiry)
         self.client.expire(self.total_key, self.key_expiry)
         self.client.delete(self.rule_cancellation_key)
         self.set_rule_initiation_key()
 
+    def mark_shard_complete(self, db_alias):
+        completed_shards = self.client.get(self.completed_shards_key) or []
+        self.client.set(self.completed_shards_key, completed_shards + ['db_alias'])
+
+    def all_shards_complete(self):
```

Comment:
nit: `all_shared_completed` would be a better name

Reply:
Sure